core/src/main/java/org/infinispan/interceptors/locking/PessimisticLockingInterceptor.java

```diff
@@ -85,21 +85,25 @@ public void init(CommandsFactory factory) {
 
    @Override
    public final Object visitGetKeyValueCommand(InvocationContext ctx, GetKeyValueCommand command) throws Throwable {
+      boolean releaseRemote = false;
       try {
          if (command.hasFlag(Flag.FORCE_WRITE_LOCK) && ctx.isInTxScope()) {
             boolean skipLocking = hasSkipLocking(command);
             if (!skipLocking) {
+               releaseRemote = true;
                acquireRemoteIfNeeded(ctx, command, cdl.localNodeIsPrimaryOwner(command.getKey()));
             }
             long lockTimeout = getLockAcquisitionTimeout(command, skipLocking);
             lockKeyAndCheckOwnership(ctx, command.getKey(), lockTimeout, skipLocking);
          }
          return invokeNextInterceptor(ctx, command);
       } catch (Throwable t) {
+         releaseRemote = false;
          releaseLocksOnFailureBeforePrepare(ctx);
          throw t;
       } finally {
          if (!ctx.isInTxScope()) lockManager.unlockAll(ctx);
+         if(releaseRemote) releaseRemoteIfNeeded(ctx, command, cdl.localNodeIsPrimaryOwner(command.getKey()));
```

Comment:
This seems wrong, the locks should only be released on commit/rollback. If we release the lock immediately after reading it, then we might as well not lock it at all, like we do when the `FORCE_WRITE_LOCK` flag isn't used.